internal/crypto/ptls/default.go

```diff
@@ -57,16 +192,54 @@ func Default(rootCAs *x509.CertPool) *tls.Config {
 	}
 }
 
+// buildCipherSuites returns the intersection of its parameters, as a list of cipher suite IDs.
+// If allowedCipherSuites is empty, it will return all the configuredCipherSuites.
+func buildCipherSuites(
+	configuredCipherSuites []*tls.CipherSuite,
+	allowedCipherSuites []*tls.CipherSuite,
+) []uint16 {
+	allowedCipherSuiteIDs := make([]uint16, 0)
+
+	for _, allowedCipherSuite := range allowedCipherSuites {
+		for _, secureCipher := range configuredCipherSuites {
+			if secureCipher.ID == allowedCipherSuite.ID {
+				allowedCipherSuiteIDs = append(allowedCipherSuiteIDs, secureCipher.ID)
+			}
+		}
+	}
+
+	configuredCipherSuiteIDs := make([]uint16, len(configuredCipherSuites))
+	for i := range configuredCipherSuites {
+		configuredCipherSuiteIDs[i] = configuredCipherSuites[i].ID
+	}
+
+	// If the user did not provide any valid allowed cipher suites, use the configured allowed cipher suites.
+	// Note that the allowed cipher suites are validated elsewhere, so this should only happen when the user chose not to specify any allowed cipher suites.
+	if len(allowedCipherSuiteIDs) == 0 {
+		allowedCipherSuiteIDs = configuredCipherSuiteIDs
+	}
+	// Preserve the order as shown in configuredCipherSuites
+	slices.SortFunc(allowedCipherSuiteIDs, func(a, b uint16) int {
+		return slices.Index(configuredCipherSuiteIDs, a) - slices.Index(configuredCipherSuiteIDs, b)
+	})
+
+	return allowedCipherSuiteIDs
+}
+
+// Default or Secure should be the only way to create a tls.Config within any component of Pinniped.
+func Default(rootCAs *x509.CertPool) *tls.Config {
+	return buildTLSConfig(rootCAs, cipherSuitesForDefault(), getAllowedCiphersForTLSOneDotTwo())
+}
+
+// DefaultLDAP or Secure should be the only ways to create a tls.Config for LDAP within any component of Pinniped.
 func DefaultLDAP(rootCAs *x509.CertPool) *tls.Config {
-	c := Default(rootCAs)
-	// add less secure ciphers to support the default AWS Active Directory config
-	c.CipherSuites = append(c.CipherSuites,
-		// CBC with ECDHE
-		// this provides forward secrecy and confidentiality of data but not authenticity
-		// MAC-then-Encrypt CBC ciphers are susceptible to padding oracle attacks
-		// See https://crypto.stackexchange.com/a/205 and https://crypto.stackexchange.com/a/224
-		tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
-		tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
-	)
-	return c
+	return buildTLSConfig(rootCAs, cipherSuitesForDefaultLDAP(), getAllowedCiphersForTLSOneDotTwo())
```

Comment:
Might it make sense to pre-compute the tls.Config at configuration time, instead of recomputing it every time someone calls `Default` or `DefaultLDAP`?

Reply:
I did think about that. I didn't go that route just due to the difficulty of synchronizing and test pollution and stuff like that.